pkg/serializer/func_queue.go

```diff
@@ -3,100 +3,52 @@
 
 package serializer
 
-import (
-	"context"
-	"fmt"
-)
-
-var (
-	// NoRetry always returns false independently of the number of retries.
-	NoRetry = func(int) bool { return false }
-)
-
-// WaitFunc will be invoked each time a queued function has returned an error.
-// nRetries will be set to the number of consecutive execution failures that
-// have occurred so far. The WaitFunc must return true if execution must be
-// retried or false if the function must be returned from the queue.
-type WaitFunc func(nRetries int) bool
-
-type queuedFunction struct {
-	f        func() error
-	waitFunc WaitFunc
-}
-
 type FunctionQueue struct {
-	queue  chan queuedFunction
+	queue  chan func() error
 	stopCh chan struct{}
+	err    error
 }
 
 // NewFunctionQueue returns a FunctionQueue that will be used to execute
 // functions in the same order they are enqueued.
 func NewFunctionQueue(queueSize uint) *FunctionQueue {
 	fq := &FunctionQueue{
-		queue:  make(chan queuedFunction, queueSize),
+		queue:  make(chan func() error),
 		stopCh: make(chan struct{}),
 	}
 	go fq.run()
 	return fq
 }
 
-// run starts the FunctionQueue internal worker. It will be stopped once
-// `stopCh` is closed or receives a value.
+// run the queue's internal worker. Returns when stopCh is closed or when
+// a function has been dequeued and executed. Closes stopCh after invoking
+// a function.
 func (fq *FunctionQueue) run() {
-	for {
-		select {
-		case <-fq.stopCh:
-			return
-		case f := <-fq.queue:
-			retries := 0
-			for {
-				select {
-				case <-fq.stopCh:
-					return
-				default:
-				}
-				retries++
-				if err := f.f(); err != nil {
-					if !f.waitFunc(retries) {
-						break
-					}
-				} else {
-					break
-				}
-			}
-		}
+	select {
+	case f := <-fq.queue:
+		fq.err = f()
+		// Unblock all callers to Wait().
+		close(fq.stopCh)
+	case <-fq.stopCh:
+		return
 	}
 }
 
-// Stop stops the function queue from processing the functions on the queue.
-// If there are functions in the queue waiting for them to be processed, they
-// won't be executed.
-func (fq *FunctionQueue) Stop() {
-	close(fq.stopCh)
+// Wait for the queue to be stopped.
+//
+// Returns any error returned by an enqueued function.
+func (fq *FunctionQueue) Wait() error {
+	<-fq.stopCh
+	return fq.err
 }
 
-// Wait until the FunctionQueue is stopped, or the specified context deadline
-// expires. Returns the error from the context, or nil if the FunctionQueue
-// was completed before the context deadline.
-func (fq *FunctionQueue) Wait(ctx context.Context) error {
+// Enqueue f to the queue. Blocks if the queue is full.
+// Returns immediately without enqueueing if the queue has been closed.
+func (fq *FunctionQueue) Enqueue(f func() error) {
 	select {
 	case <-fq.stopCh:
-	case <-ctx.Done():
-	}
-	if err := ctx.Err(); err != nil {
-		return fmt.Errorf("serializer: %w", err)
+		return
+	default:
```

Comment:
It looks to me like this last commit refactors a bunch of this datastructure, but the core functional change here would be the same as if `Enqueue` just stored the `err` and `Wait(...)` returned that err instead of returning `nil` at the end of the function.

Reply:
Correct, this is mentioned in the commit message and subject. Just an observation, or changes required?

Reply:
Yeah just an observation. I struggled quite a bit with this patch during the previous round of review because of how it refactored the entire datastructure and the users rather than making the small functional change described above then doing the datastructure rewrite second. There are ways to split up such changes into smaller, more easily reviewable chunks.